File: backend/src/main/java/com/proyecto3d/backend/apirest/model/service/CarritoService.java
package com.proyecto3d.backend.apirest.model.service;

import java.util.List;
import java.util.Optional;

import com.proyecto3d.backend.apirest.model.dto.CarritoDTO;
import com.proyecto3d.backend.apirest.model.entity.Carrito;

/**
 * Interfaz del servicio para la gestión del carrito
 */
public interface CarritoService {

    /**
     * Añade un elemento al carrito
     * @param usuarioId ID del usuario
     * @param anuncioId ID del anuncio
     * @param cantidad Cantidad a añadir
     * @param materialSeleccionado Material seleccionado
     * @param colorSeleccionado Color seleccionado (opcional)
     * @param acabadoPremium Si incluye acabado premium
     * @param urgente Si es urgente
     * @param envioGratis Si incluye envío gratis
     * @return El elemento del carrito creado o actualizado
     */
    CarritoDTO anadirAlCarrito(Long usuarioId, Long anuncioId, Integer cantidad, 
                             String materialSeleccionado, String colorSeleccionado,
                             Boolean acabadoPremium, Boolean urgente, Boolean envioGratis);

    /**
     * Obtiene todos los elementos del carrito de un usuario
     * @param usuarioId ID del usuario
     * @return Lista de DTOs del carrito
     */
    List<CarritoDTO> obtenerCarritoUsuario(Long usuarioId);

    /**
     * Actualiza la cantidad de un elemento del carrito
     * @param carritoId ID del elemento del carrito
     * @param nuevaCantidad Nueva cantidad
     * @return El elemento del carrito actualizado
     */
    CarritoDTO actualizarCantidad(Long carritoId, Integer nuevaCantidad);

    /**
     * Actualiza los servicios adicionales de un elemento del carrito
     * @param carritoId ID del elemento del carrito
     * @param acabadoPremium Si incluye acabado premium
     * @param urgente Si es urgente
     * @param envioGratis Si incluye envío gratis
     * @return El elemento del carrito actualizado
     */
    CarritoDTO actualizarServiciosAdicionales(Long carritoId, Boolean acabadoPremium, 
                                            Boolean urgente, Boolean envioGratis);

    /**
     * Elimina un elemento del carrito
     * @param carritoId ID del elemento del carrito
     */
    void eliminarElementoCarrito(Long carritoId);

    /**
     * Vacía todo el carrito de un usuario
     * @param usuarioId ID del usuario
     */
    void vaciarCarrito(Long usuarioId);

    /**
     * Obtiene el número de elementos en el carrito de un usuario
     * @param usuarioId ID del usuario
     * @return Número de elementos
     */
    Long contarElementosCarrito(Long usuarioId);

    /**
     * Calcula el precio total del carrito de un usuario
     * @param usuarioId ID del usuario
     * @return Precio total del carrito
     */
    Double calcularTotalCarrito(Long usuarioId);
} 

File: frontend/src/app/paginas/carrito/carrito.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface CarritoElemento {
  id: number;
  usuarioId: number;
  anuncioId: number;
  tituloAnuncio: string;
  descripcionAnuncio: string;
  imagenAnuncio: string;
  nombreProveedor: string;
  cantidad: number;
  materialSeleccionado: string;
  colorSeleccionado?: string;
  acabadoPremium: boolean;
  urgente: boolean;
  envioGratis: boolean;
  precioUnitario: number;
  precioTotal: number;
  fechaAgregado: string;
}

export interface CarritoResponse {
  carrito: CarritoElemento[];
  totalElementos: number;
  totalPrecio: number;
}

export interface AnadirCarritoRequest {
  usuarioId: number;
  anuncioId: number;
  cantidad: number;
  materialSeleccionado: string;
  colorSeleccionado?: string;
  acabadoPremium: boolean;
  urgente: boolean;
  envioGratis: boolean;
}

@Injectable({
  providedIn: 'root'
})
export class CarritoService {

  private apiUrl = 'http://localhost:8080/api/carrito';

  private httpOptions = {
    headers: new HttpHeaders({
      'Content-Type': 'application/json'
    })
  };

  constructor(private http: HttpClient) { }

  /**
   * Añade un elemento al carrito
   */
  anadirAlCarrito(request: AnadirCarritoRequest): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/anadir`, request, this.httpOptions);
  }

  /**
   * Obtiene el carrito de un usuario
   */
  obtenerCarritoUsuario(usuarioId: number): Observable<CarritoResponse> {
    return this.http.get<CarritoResponse>(`${this.apiUrl}/usuario/${usuarioId}`);
  }

  /**
   * Actualiza la cantidad de un elemento del carrito
   */
  actualizarCantidad(carritoId: number, nuevaCantidad: number): Observable<any> {
    return this.http.put<any>(`${this.apiUrl}/${carritoId}/cantidad?nuevaCantidad=${nuevaCantidad}`, {});
  }

  /**
   * Actualiza los servicios adicionales de un elemento del carrito
   */
  actualizarServiciosAdicionales(carritoId: number, servicios: {
    acabadoPremium: boolean;
    urgente: boolean;
    envioGratis: boolean;
  }): Observable<any> {
    return this.http.put<any>(`${this.apiUrl}/${carritoId}/servicios`, servicios, this.httpOptions);
  }

  /**
   * Elimina un elemento del carrito
   */
  eliminarElementoCarrito(carritoId: number): Observable<any> {
    return this.http.delete<any>(`${this.apiUrl}/${carritoId}`);
  }

  /**
   * Vacía todo el carrito de un usuario
   */
  vaciarCarrito(usuarioId: number): Observable<any> {
    return this.http.delete<any>(`${this.apiUrl}/vaciar/${usuarioId}`);
  }

  /**
   * Obtiene el número de elementos en el carrito de un usuario
   */
  contarElementosCarrito(usuarioId: number): Observable<{count: number}> {
    return this.http.get<{count: number}>(`${this.apiUrl}/count/${usuarioId}`);
  }

  /**
   * Calcula el precio total del carrito de un usuario
   */
  calcularTotalCarrito(usuarioId: number): Observable<{total: number}> {
    return this.http.get<{total: number}>(`${this.apiUrl}/total/${usuarioId}`);
  }
} 